### PR TITLE
try shrink heap at some key points

### PR DESCRIPTION
**Pull request by zyfjeff**
Signed-off-by: tianqian.zyf <tianqian.zyf@alibaba-inc.com>

  The purpose of this pr is to release the cache introduced by TCmalloc,
  mainly in xDS config updates, admin handler, and so on. all work on the main thread,
  so the overall impact on performance is small.

  Ref: https://github.com/envoyproxy/envoy/pull/9471#discussion_r363825985

Description: try to release the cache introduced by TCmalloc,
Risk Level: low
Testing: N/A
Docs Changes: N/A
Release Notes:
[Optional Fixes #Issue] #9381 #9471 
[Optional Deprecated:]

### CODE REVIEW

File: source/common/memory/utils.cc [hunk 1]
```diff
@@ -13,5 +16,26 @@ void Utils::releaseFreeMemory() {
...
+  mainly in xDS config updates, admin handler, and so on. all work on the main thread,
+  so the overall impact on performance is small.
+  Ref: https://github.com/envoyproxy/envoy/pull/9471#discussion_r363825985
+*/
+void Utils::tryShrinkHeap() {
+#ifdef TCMALLOC
+  // TODO(zyfjeff): Make max unfreed memory byte configurable
+  static const uint64_t MAX_UNFREED_MEMORY_BYTE = 100 * 1024 * 1024;
+  auto heap_size = Stats::totalCurrentlyReserved() - Stats::totalPageHeapUnmapped();
+  auto allocate_size = Stats::totalCurrentlyAllocated();
```
Comment (yanavlasov) [inline source/common/memory/utils.cc]: nit: s/allocate_size/allocated_size

File: source/common/memory/utils.cc [hunk 2]
```diff
@@ -13,5 +16,26 @@ void Utils::releaseFreeMemory() {
...
+  The purpose of this function is to release the cache introduced by tcmalloc,
+  mainly in xDS config updates, admin handler, and so on. all work on the main thread,
+  so the overall impact on performance is small.
+  Ref: https://github.com/envoyproxy/envoy/pull/9471#discussion_r363825985
+*/
+void Utils::tryShrinkHeap() {
+#ifdef TCMALLOC
+  // TODO(zyfjeff): Make max unfreed memory byte configurable
+  static const uint64_t MAX_UNFREED_MEMORY_BYTE = 100 * 1024 * 1024;
+  auto heap_size = Stats::totalCurrentlyReserved() - Stats::totalPageHeapUnmapped();
```
Comment (yanavlasov) [inline source/common/memory/utils.cc]: Maybe for slightly more clarify call it `committed_heap_size`? I.e. heap size that has a physical page mapped. Also can you add a comment why you want to exclude unmapped pages from this computation, please?

Comment (zyfjeff) [reply source/common/memory/utils.cc]: Yes, it's a little confusing here, so I'll change it to 'generic.total_physical_bytes'

https://github.com/gperftools/gperftools/blob/b85652bf268d061a0ad0aec90f0d3c29aa80a6b5/src/gperftools/malloc_extension.h#L171

https://github.com/gperftools/gperftools/blob/b85652bf268d061a0ad0aec90f0d3c29aa80a6b5/src/gperftools/malloc_extension.h#L171

**Comment by zyfjeff** [comment]
**Receive the fully configured memory state**

```
# curl http://127.0.0.1:15002/memory
{
 "allocated": "1058294160",
 "heap_size": "1472217088",
 "pageheap_unmapped": "0",
 "pageheap_free": "386646016",
 "total_thread_cache": "9461896"
}
```

**Performs a ` curl http://127.0.0.1:15002/cluster ` command after the state of the memory**

```
# curl http://127.0.0.1:15002/memory
{
 "allocated": "1058303856",
 "heap_size": "3598794752",
 "pageheap_unmapped": "155648",
 "pageheap_free": "2516238336",
 "total_thread_cache": "4357856"
}
```

**Check the memory state again to make sure the memory has been reclaimed**
```
# curl http://127.0.0.1:15002/memory
{
 "allocated": "1058303856",
 "heap_size": "3598794752",
 "pageheap_unmapped": "2516393984",
 "pageheap_free": "0",
 "total_thread_cache": "4321600"
}
```

**Pull request merged**